feat: adds support for oidc publish

reggi opened the pull request:
# 🎉 Introducing OIDC Support for npm Publishing! 

> Further discussions not related directly this PR should happen here https://github.com/orgs/community/discussions/161015

We're thrilled to announce a new security feature that makes publishing npm packages from CI environments easier and more secure! This PR adds OpenID Connect (OIDC) token support for npm publishing, which eliminates the need to store long-lived access tokens in CI secrets.

With OIDC support, you can now publish packages from GitHub Actions and GitLab CI with improved security through short-lived, automatically generated tokens. This is a major step forward in securing the npm ecosystem and simplifying CI/CD workflows.

## Technical Details

This implementation adds OIDC token support by:
1. Detecting when npm is running in a supported CI environment (currently GitHub Actions and GitLab CI)
2. Retrieving an OIDC token from the CI provider
3. Exchanging this token with the npm registry for a short-lived publishing token
4. Using this token for authentication during the publish process

The feature is designed to be non-invasive - it only activates in CI environments and gracefully falls back to traditional authentication methods when OIDC isn't available.

## For Publishers

### Updating Package Settings on npmjs.com
> [!WARNING]  
> Not live yet, OIDC support is currently under development. The CLI, `npmjs.com`, and registry changes will be rolled out incrementally. Stay tuned for a public preview announcement. As of now, this documentation reflects features planned for a future release.

> [!IMPORTANT]  
> In order to use OIDC publishing, a package must already exist on npmjs.com. This means the initial publish needs to be done through conventional means; further publishes, once configured, can use OIDC.

Before using OIDC for publishing, ensure your package settings on npmjs.com are configured to allow CI/CD workflows. 
1. Log in to your npm account and navigate to the **Package Settings** page for the package you want to publish.
2. Add a new Connection for a Trusted Publisher and fill out the details.
3. Save your changes to apply the new settings.

This step ensures that your package is ready to accept tokens generated via OIDC workflows.

### GitHub Actions

To publish with OIDC from GitHub Actions:

1. Add the `id-token: write` permission to your workflow:

```yaml
permissions:
  id-token: write
  contents: read
```

2. No need to set up `NPM_TOKEN` secrets anymore! Just run npm publish as usual:

```yaml
- name: Publish to npm
  run: npm publish
```

### GitLab CI

To publish with OIDC from GitLab CI:

1. Configure your GitLab CI pipeline to provide the ID token via the `NPM_ID_TOKEN` environment variable:

```yaml
publish:
  script:
    - npm publish
  id_tokens:
    NPM_ID_TOKEN:
      aud: npm:registry.npmjs.org
```

## For other CLI's

If you're building a CLI tool that publishes to npm registries, you can implement OIDC support by:

1. Detecting CI environments
2. For GitHub Actions, requesting tokens via the `ACTIONS_ID_TOKEN_REQUEST_URL` endpoint with the proper audience format (`npm:registry.hostname`)
3. For GitLab CI, using the `NPM_ID_TOKEN` environment variable if available
4. Exchanging the OIDC token with the npm registry's token exchange endpoint

## For other Registries

As a registry, you'll need a way for package publishers to create connections between OIDC Trusted Publishers and the registry, similar to how we allow connections to be added on the package settings page of npmjs.com.

To support OIDC token authentication in your npm-compatible registry:

1. Implement an endpoint at `/-/npm/v1/oidc/token/exchange/package/${escapedPackageName}` that accepts POST requests (no body) with `Authorization` header / `Bearer` set to the `jwt-token-from-ci-provider`. 

2. Verify the OIDC token using standard JWT validation practices, checking the audience claim matches your expected format (`npm:your.registry.hostname`)

3. Return a response with a short-lived npm token:
   ```json
   {
     "token": "npm_short_lived_token"
   }
   ```

--- 

# Technical Overview

* The OIDC integration begins in the publish command.
* This adds a new utility module `oidc.js` handles:
* Full test coverage has been implemented in the publish test suite. ✅ 
* The authentication flow follows this path:
  * `publish` command → `libnpmpublish` module → `npm-registry-fetch`

Key touchpoints:
* https://github.com/npm/cli/blob/latest/lib/commands/publish.js#L184
* https://github.com/npm/cli/blob/latest/workspaces/libnpmpublish/lib/publish.js#L53-L63
* https://github.com/npm/npm-registry-fetch/blob/main/lib/index.js#L55

This initial implementation is focused on the publish workflow only. Currently OIDC token support is limited to the publish command.

ljharb commented:
1. Is there a config option or command line argument that can *prevent* OIDC from kicking in? (it seems like "write" is a magic id-token string for GHA, so i'm not sure what i'd put there or in an env var to guarantee that it can't kick in)
1. if i wanted to publish using OIDC on my local machine, for testing or for funsies, what would I need to set up and how would I need to invoke it?

reggi commented:
@ljharb 👋 

> Is there a config option or command line argument that can prevent OIDC from kicking in? (It seems like "write" is a magic id-token string for GHA, so I'm not sure what I'd put there or in an env var to guarantee that it can't kick in.)

I've thought about this a lot. I also want this, but I question its necessity. If you have `id: write` and don't configure the trusted connection on `npmjs.com`, the token-exchange endpoint will return a 404 and you won't get a token, and the publish flow will continue as normal. So, you're already opted out by not configuring it—though, yes, there is still an extra network request.

> If I wanted to publish using OIDC on my local machine, for testing or for fun, what would I need to set up and how would I need to invoke it?

Not really possible. The whole point of OIDC is to have a trusted publisher (like GitLab or GitHub) as the issuer of a token that the registry trusts. A local machine isn't a trusted issuer, so we wouldn't be able to validate any token you could provide.

ljharb commented:
Right, but what if an attacker configures OIDC on npmjs.com unbeknownst to me? I'd still want to ensure CI can't publish with it.

So to clarify, the reason it's not possible is because the npm servers only have a finite hardcoded list of "trusted OIDC publishers", and i'm not on it?

wraithgar commented on lib/utils/oidc.js:
```diff
@@ -0,0 +1,138 @@
+      ciInfo.GITHUB_ACTIONS,
+      /** @see https://github.com/watson/ci-info/blob/v4.2.0/vendors.json#L161C13-L161C22 */
+      ciInfo.GITLAB,
+    ].some((env) => env))) {
```
```suggestion
    ].includes(env)) {
```

wraithgar commented on lib/utils/oidc.js:
```diff
@@ -0,0 +1,138 @@
+     * Additionally, the token is required by the "live" configuration or getters within `config`.
+     */
+    opts[authTokenKey] = response.token
+    config.set(authTokenKey, response.token, 'user')
```
Oooh this seems dangerous.  I don't think we have any other code tha

wraithgar commented on lib/utils/oidc.js:
```diff
@@ -0,0 +1,138 @@
+     * Additionally, the token is required by the "live" configuration or getters within `config`.
+     */
+    opts[authTokenKey] = response.token
+    config.set(authTokenKey, response.token, 'user')
```
This seems like a code smell.  Either we're using config or we're using flatOptions.  It seems we're trying to do both?

wraithgar replied:
The TODOs at the top of `lib/index.js` in the config workspace speak directly to this.

wraithgar reviewed (dismissed):
So clean, so boilerplate.  Let's leave this unmerged for at least a little while to see if any more community input happens.

reggi commented:
@ljharb 

> Right, but what if an attacker configures OIDC on npmjs.com unbeknownst to me? I'd still want to ensure CI can't publish with it.

If an attacker has access to your npmjs.com account, they can just make gat tokens, I think OIDC would be the least of your worries 🤔 . Further expanding on this idea if a malicious actor got access to your account and added a trusted connection to one of THEIR repos, with their own workflow a `--no-oidc` flag isn't gonna help you at that point, they write the workflow, and the `npm publish` script in the workflow, they can just omit that flag.

> So to clarify, the reason it's not possible is because the npm servers only have a finite hardcoded list of "trusted OIDC publishers", and i'm not on it?

Yeah, kinda

reggi commented:
Further expanding on this idea if a malicious actor got access to your account and added a trusted connection to one of THEIR repo, with their own workflow a `--no-oidc` flag isn't gonna help you at that point, they write the workflow, and the `npm publish` script in the workflow.

leobalter commented:
Just to chime in, I'm onboard and supportive w/ the decision to avoid a cli parameter here for the sake of feature streamlining. Security wise, the malicious access goes similarly to creating access tokens, as mentioned above.

> So to clarify, the reason it's not possible is because the npm servers only have a finite hardcoded list of "trusted OIDC publishers", and i'm not on it?

By finite we are kick starting this feature w/ **2 trusted publishers**: GitHub Actions and GitLab.

This is an MVP and we want to understand the usage feedback for the next iterations on this feature.

---

If possible, I'd really love if we can shift other feature requests and feedback at the [community discussion](https://github.com/orgs/community/discussions/161015) post as it makes it easier for me to track feedback received and we keep this PR limited to new changes added to the repo.

hashtagchris commented on lib/utils/oidc.js:
```diff
@@ -0,0 +1,147 @@
+      return undefined
+    }
+
+    log.silly('oidc', 'Detemrmining if npm should use OIDC publishing')
```
> Detemrmining

typo

hashtagchris replied:
```suggestion
    log.silly('oidc', 'Determining if npm should use OIDC publishing')
```

hashtagchris commented on lib/utils/oidc.js:
```diff
@@ -0,0 +1,147 @@
+          const url = new URL(process.env.ACTIONS_ID_TOKEN_REQUEST_URL)
+          url.searchParams.append('audience', audience)
+          const response = await fetch(url.href, {
+            method: 'POST',
```
`POST` seems reasonable to me, given that the endpoint mints a token. However [GitHub's curl example](https://docs.github.com/en/actions/security-for-github-actions/security-hardening-your-deployments/about-security-hardening-with-openid-connect#updating-your-actions-for-oidc) suggests `GET` should be used, as does Actions code:

https://github.com/actions/toolkit/blob/683703c1149439530dcee7b8c5dbbfeec4104368/packages/core/src/oidc-utils.ts#L46-L50

https://github.com/github/actions-oidc-debugger/blob/0705a46041e35206455b781a260e4af1c6e60b6d/actionsoidc/actions-oidc.go#L86

```suggestion
            method: 'GET',
```

hashtagchris commented on lib/commands/publish.js:
```diff
@@ -136,6 +137,9 @@ class Publish extends BaseCommand {
 
     const registry = npmFetch.pickRegistry(resolved, opts)
+
+    await oidc({ packageName: manifest.name, registry, opts, config: this.npm.config })
```
Optional: Is there a pithy way we can indicate we're doing an OIDC publish in the the `notice` level log output? One idea:

```diff
 npm notice Tarball Contents
 npm notice 167B README.md
 npm notice 236B package.json
...
 npm notice total files: 2
 npm notice
- npm notice Publishing to https://registry.npmjs.org/ with tag latest and default access
+ npm notice Publishing to https://registry.npmjs.org/ with tag latest and default access, using OIDC
```

I would only include `using OIDC` or `as {CI Provider Name}` if the token exchange was successful and we're overriding the auth token.

reggi replied:
There's no way the system currently knows that it's using OIDC (no global variables or state to use where that log resides), we can just have one log that says "Using OIDC" within the OIDC file flow as an alternative. I'd rather not like to maintain state.

hashtagchris commented on lib/utils/oidc.js:
```diff
@@ -0,0 +1,146 @@
+          log.silly('oidc:', 'GITHUB_ACTIONS valid fetch response for id_token')
+          idToken = json.value
+        } else {
+          throw new Error('GITHUB_ACTIONS detected. If you intend to publish using OIDC, please set workflow permissions for `id-token: write`')
```
nit: This doesn't strike me as an error condition

```suggestion
          log.verbose('oidc', 'GITHUB_ACTIONS detected. If you intend to publish using OIDC, please set workflow permissions for `id-token: write`')
          return undefined
```

reggi replied:
My rationale is that anything that is not a successful retrieval of a token, throws, as a method of exiting the flow, rather than returning, and the error is caught and logged at the bottom.

hashtagchris replied:
By that rationale, should lines 33 and 116 throw instead of returning early? Both are within the `try` block.

reggi committed:
feat: adds support for oidc publish

reggi committed:
fix: change to the oidc flow for more granular control over log levels (#8399)

Co-authored-by: Chris Sidi <hashtagchris@github.com>

reggi committed:
Small change to registry fetch options

I don't believe we need to destructure the objects as we we're doing.

jakebailey commented:
I feel like this logic should not be in the npm CLI, but rather a part of `setup-node` or some other third party lib. Otherwise, this code will end up continuing to include more and more special cases for more CI providers, and have to be reimplemented by every other package manager or tool that needs the npm token directly (like the DefinitelyTyped infrastructure that publishes directly via the API, which I would love to use this with).

hashtagchris commented on lib/utils/oidc.js:
```diff
@@ -0,0 +1,169 @@
+    const regKey = `//${parsedRegistry.host}${parsedRegistry.pathname}`
+    const authTokenKey = `${regKey}:_authToken`
+
+    const exitingToken = config.get(authTokenKey)
```
```suggestion
    const existingToken = config.get(authTokenKey)
```

hashtagchris commented on lib/utils/oidc.js:
```diff
@@ -0,0 +1,169 @@
+    const authTokenKey = `${regKey}:_authToken`
+
+    const exitingToken = config.get(authTokenKey)
+    if (exitingToken) {
+      log.silly('oidc', 'Existing token found')
+    } else {
+      log.silly('oidc', 'No existing token found')
+    }
```
Is the check for an existing token config useful? Maybe we should simplify and remove this code.

reggi replied:
I do think it's useful to know if we're about to potentially overwrite an existing token or not, but i have no strong feeling about it. If someone sees `Existing token found` in the workflow, it can also remind them to remove that token if they only use it to publish.

reggi committed:
Merge branch 'latest' into oidc

reggi committed:
feat: npm view oidc considerations (#8432)

reggi committed:
feat: oidc provenance by default (#8412)

This PR adds "auto" or "default" provenance to publishes that use OIDC
within github and gitlab. It does this by checking the OIDC id token
payload and checking if the current repo's visibility is public or
private if it's public we do the equivalent of adding the `--provenance`
flag.

reggi committed:
give it a dry-run test

reggi committed:
fix: checks for sigstore env for gitlab provenance (#8439)

small update to allow gitlab to NOT REQUIRE provenance

owlstronaut reviewed (dismissed):
This looks great. The documentation and code in oidc.js reads like a well laid out story with an easy to follow plot 📖

reggi committed:
chore: simplifies the oidc test / nocks (#8446)

reggi committed:
rm exisiting token check

reggi committed:
updated logs

reggi committed:
normalize the verbiage

reggi committed:
chore: cleanup logs (#8451)

reggi committed:
chore: remove unnecessary ifs (#8453)

reggi committed:
last minute decision to not log anything locally
